src/registrar/models/models.py

```diff
@@ -0,0 +1,81 @@
+from django.contrib.auth.models import AbstractUser
+from django.db import models
+
+
+class User(AbstractUser):
+    """
+    A custom user model that performs identically to the default user model
+    but can be customized later.
+    """
+
+    def __str__(self):
+        if self.userprofile.display_name:
+            return self.userprofile.display_name
+        else:
+            return self.username
+
+
+class TimeStampedModel(models.Model):
+    """
+    An abstract base model that provides self-updating
+    `created_at` and `updated_at` fields.
+    """
+
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+
+    class Meta:
+        abstract = True
+        # don't put anything else here, it will be ignored
+
+
+class AddressModel(models.Model):
+    """
+    An abstract base model that provides common fields
+    for postal addresses.
+    """
+
+    # contact's street (null ok)
+    street1 = models.TextField(blank=True)
+    # contact's street (null ok)
+    street2 = models.TextField(blank=True)
+    # contact's street (null ok)
+    street3 = models.TextField(blank=True)
+    # contact's city
+    city = models.TextField(blank=True)
+    # contact's state or province (null ok)
+    sp = models.TextField(blank=True)
+    # contact's postal code (null ok)
+    pc = models.TextField(blank=True)
+    # contact's country code
+    cc = models.TextField(blank=True)
+
+    class Meta:
+        abstract = True
+        # don't put anything else here, it will be ignored
+
+
+class ContactModel(models.Model):
+    """
+    An abstract base model that provides common fields
+    for contact information.
+    """
+
+    voice = models.TextField(blank=True)
+    fax = models.TextField(blank=True)
+    email = models.TextField(blank=True)
+
+    class Meta:
+        abstract = True
+        # don't put anything else here, it will be ignored
+
+
+class UserProfile(TimeStampedModel, ContactModel, AddressModel):
+    user = models.OneToOneField(User, null=True, on_delete=models.CASCADE)
```

Comment:
Is the `on_delete` attribute uni-directional for a `OneToOneField`? (Deleting UserProfile deletes User, but deleting User does _not_ delete UserProfile.) I assume so, but want to confirm.

Reply:
It is uni-directional, but it goes the opposite of the direction that you specify. Even though we are specifying it here on the `UserProfile` model, `on_delete` describes what should happen when we delete the matching `User`. 

In this case, a `User` who is deleted no longer needs a `UserProfile` so we want to also remove the `UserProfile` at that time which is what `CASCADE` indicates.

https://docs.djangoproject.com/en/4.1/ref/models/fields/#django.db.models.ForeignKey.on_delete for the formal documentation which just refers to SQL's behavior and https://stackoverflow.com/a/38389488 for a better explanation.

Reply:
Ack, thank you for the docs link! I do tend to get confused by foreign key constraints since the key is so often held in the table which is opposite to the "natural" relationship (e.g. a car has wheels, but a Wheel has the key to the Car...)

I have a [draft ADR](https://docs.google.com/document/d/1ea04AvWeXwxelxmhD2rqMfcRBnKdnxvam6EyA8ehe7I/edit#bookmark=id.po74ml55c5lk) for the business logic decisions that are going into the design of the database models (e.g. "a User who is deleted no longer needs a UserProfile"), but it is kinda messy and has met with some objections from the team. Would you be good enough to start a new one in that document and include your thinking behind the models on this PR?